### PR TITLE
Implement full DkgSign over StackerDB

### PR DESCRIPTION
**Pull request by xoloki**
### Description

This PR builds on the `test_stackerdb_dkg` integration test, but now it uses  `libstackerdb` and `libsigner`.  It implements a full state-machine based `stacks-signer`, with all of the crypto and networking necessary to run a full `DkgSign` command.

### Applicable issues
- Closes https://github.com/stacks-network/stacks-blockchain/issues/3881

### Checklist
- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

### CODE REVIEW

File: libsigner/src/runloop.rs
Comment (jcnelson) [inline libsigner/src/runloop.rs:90]: Before merging, can you remove the commented-out debug statements?

File: stacks-signer/src/config.rs
Comment (jcnelson) [inline stacks-signer/src/config.rs:99]: I'm having a hard time understanding the difference between `message_private_key` and `stacks_private_key`.  Can you help me understand why there need to be two private keys?

File: stacks-signer/src/crypto/frost.rs
Comment (jcnelson) [inline stacks-signer/src/crypto/frost.rs]: Does this need to be addressed in order for the code to be robust against a malicious signer?

File: stacks-signer/src/crypto/frost.rs
Comment (jcnelson) [inline stacks-signer/src/crypto/frost.rs:429]: From the demo today, it seemed that this happened?  Can this happen as part of the signer's normal operation?  If so, what would the signer need to do here?

File: stacks-signer/src/crypto/frost.rs
Comment (jcnelson) [inline stacks-signer/src/crypto/frost.rs:472]: I love how this system checks for a valid state transition!  Do you think it makes sense to outright `panic!()` if we encounter an invalid state transaction? Because, that would indicate a logic bug in the signer, right?

File: stacks-signer/src/crypto/frost.rs
Comment (jcnelson) [inline stacks-signer/src/crypto/frost.rs:87]: It seems these fields (`dkg_public_shares`, `public_nonces`, `signature_shares`, `aggregate_public_key`, `signature`, and `schnorr_proof`) are generated as part of either DKG or a signing round.  If so, then would it make sense to treat them as `Option<..>`, so it's unambiguous as to whether or not they have been calculated as part of this DKG or signing round?

File: stacks-signer/src/crypto/frost.rs
Comment (jcnelson) [inline stacks-signer/src/crypto/frost.rs:367]: Will this TODO be addressed before merge?

File: stacks-signer/src/runloop.rs
Comment (jcnelson) [inline stacks-signer/src/runloop.rs:31]: Presumably, the `message` will be decoded and validated by the business logic before signing, right?  For example, the signer might want to check that the data to sign is a partially-signed Bitcoin transaction, or a well-formed Stacks block, but nothing else? If so, then would it make sense to treat this field as an `enum` of the possible types of data that could be signed, so only valid inputs can be represented?

File: stacks-signer/src/runloop.rs
Comment (jcnelson) [inline stacks-signer/src/runloop.rs:177]: Should this be `(config.signer_ids_public_keys.key_ids.len() * 7) / 10`?

File: stacks-signer/src/runloop.rs
Comment (jcnelson) [inline stacks-signer/src/runloop.rs:197]: This `- 1` can never underflow, right?

File: stacks-signer/src/runloop.rs
Comment (jcnelson) [inline stacks-signer/src/runloop.rs:230]: Flagging this TODO for completion before merge.

File: stacks-signer/src/runloop.rs
Comment (jcnelson) [inline stacks-signer/src/runloop.rs:346]: Flagging this for completion

File: testnet/stacks-node/src/event_dispatcher.rs
Comment (jcnelson) [inline testnet/stacks-node/src/event_dispatcher.rs:95]: Ha, thanks for DRY'ing this up!

File: testnet/stacks-node/src/tests/signer.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/signer.rs:37]: Could `build_contract` and `build_signer_config_tomls` be dropped here, in favor of the ones in the `stacks-signer` crate?

Comment (jcnelson) [reply testnet/stacks-node/src/tests/signer.rs:37]: Ah, I see that already happened in @jferrant's follow-on PR.

File: libsigner/src/runloop.rs
Comment (xoloki) [reply libsigner/src/runloop.rs]: Done.

File: stacks-signer/src/crypto/frost.rs
Comment (xoloki) [reply stacks-signer/src/crypto/frost.rs]: It needs to be addressed for general robustness, and specifically we will need it for the meta-protocols FIRE/ROAST (where we handle byzantine signers).  It's fine as is for now, but if you want I can do an initial robustness pass before merging this PR.

File: stacks-signer/src/crypto/frost.rs
Comment (xoloki) [reply stacks-signer/src/crypto/frost.rs:429]: We didn't see the actual warning, but we did see multiple DkgPublic and Nonce rounds.  So I think we just ran out of time for the 30 seconds the integration test allocates before stopping the Coordinator.

In a follow on PR I will add another channel so that the Signer can tell us results of operations, like the channel I added so we can give it commands async.  I can do that now if you'd like.

File: stacks-signer/src/crypto/frost.rs
Comment (xoloki) [reply stacks-signer/src/crypto/frost.rs:472]: I have a moderate distaste for `panic!` in anything that runs in a network context.  Having said that, we `panic!` if we fail to sign messages, since that means our private key is completely broken and we can't reasonably continue.  

For state machine errors, I can see a possibility of getting an unexpected message that triggers a bad state machine change.  If we can handle that kind of thing without `panic!` I think we should.

File: stacks-signer/src/crypto/frost.rs
Comment (xoloki) [reply stacks-signer/src/crypto/frost.rs:87]: I think that's a very good idea.  Let me know if you want me to do it as part of this PR or if it can wait for the follow on.

File: stacks-signer/src/crypto/frost.rs
Comment (xoloki) [reply stacks-signer/src/crypto/frost.rs:367]: Can't do it as part of this PR because the relevant code is in `stacks-sbtc`.  We are planning on number of `stacks-sbtc` changes before the follow on PR.

File: stacks-signer/src/runloop.rs
Comment (xoloki) [reply stacks-signer/src/runloop.rs:31]: Presumably there will be a small set of things we are willing to sign, and they will definitely need to be validated before doing so.  There's lots of weird crypto things that can happen if you can trick people into signing arbitrary messages.

It's definitely something to think about but seems very orthogonal to this PR itself.

File: stacks-signer/src/runloop.rs
Comment (xoloki) [reply stacks-signer/src/runloop.rs]: I certainly think so.  Hopfully clippy won't hate the extra parens.

Fixed.

File: stacks-signer/src/runloop.rs
Comment (xoloki) [reply stacks-signer/src/runloop.rs:195]: It can absolutely underflow if we don't start our `key_ids` at `1`.  But we are starting at `1`, explicitly.

After I fix `wsts` to eliminate the whole off-by-one nastiness this will go away.

File: stacks-signer/src/runloop.rs
Comment (xoloki) [reply stacks-signer/src/runloop.rs:228]: `frost-signer` is part of `stacks-sbtc`, so IMO these TODOs should be part of follow on PR after we've made a number of changes to `wsts` and `stacks-sbtc`.

File: stacks-signer/src/runloop.rs
Comment (xoloki) [reply stacks-signer/src/runloop.rs:344]: More `stacks-sbtc` changes.

File: stacks-signer/src/stacks_client.rs
Comment (xoloki) [reply stacks-signer/src/stacks_client.rs:77]: @jferrant do you want to add an issue for this, and the other TODOs and `wsts`/`stacks-sbtc` changes?

File: stacks-signer/src/stacks_client.rs
Comment (xoloki) [reply stacks-signer/src/stacks_client.rs:92]: Do you want that as part of this PR, or is it okay if @jferrant makes and issue and we deal with it later?

File: stacks-signer/src/crypto/frost.rs
Comment (xoloki) [reply stacks-signer/src/crypto/frost.rs:428]: Lol actually I had bad copypasta, the aggregate public key was actually being computed not the nonce XD

Fixed now, will test repeatedly until I see this nonce repeat and it still works.

Comment (xoloki) [reply stacks-signer/src/crypto/frost.rs:428]: Yup, just watched current code do 3 nonce rounds before getting an even Y, and SchnorrProof worked.

Fixed for realz.

File: stacks-signer/src/crypto/frost.rs
Comment (xoloki) [reply stacks-signer/src/crypto/frost.rs]: I went ahead and implemented this, which should make it less work for when I implement `FIRE`.

@jferrant we got another TODO: incrementing `dkg_public_id` if public distribution fails (due to odd y coord on agg public key) doesn't work because we don't have a `DkgPublicBegin` message; we just re-send `DkgBegin`, so we're always sending `dkg_public_id: 0` from the `SigningRound`.

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [reply stacks-signer/src/stacks_client.rs:79]: Already done. Thought I had pushed code to include the github issue? https://github.com/stacks-network/stacks-blockchain/issues/3917

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [reply stacks-signer/src/stacks_client.rs:94]: I will make an issue and link it in here. i think the sooner we get this in the better and then we can all work on the various todos without clobbering each other.

Comment (xoloki) [reply stacks-signer/src/stacks_client.rs:94]: Yeah let's try to get this merged ASAP.  I have the change ready that adds a `Sender` to everything so we can get notified whenever of all `OperationResult`s.  But it needs a tiny `wsts` fix and I don't want to hold this PR up.

File: stacks-signer/src/crypto/frost.rs
Comment (xoloki) [reply stacks-signer/src/crypto/frost.rs:488]: Follow on branch is ready with extra channel for `OperationResult`s.  Will wait until this is merged and try to batch a bunch of upstream changes.

File: stacks-signer/src/config.rs
Comment (jferrant) [reply stacks-signer/src/config.rs:99]: I can update the comment on the stacks_private_key as well to indicate what it is used for if that helps (communicating with the stacks node/stacker db contract)?

File: stacks-signer/src/crypto/frost.rs
Comment (jferrant) [reply stacks-signer/src/crypto/frost.rs:488]: Yeah adding this channel should also make the return time of the run_one_pass irrelevant and is closely related to this issue. https://github.com/stacks-network/stacks-blockchain/issues/3916

Comment (xoloki) [reply stacks-signer/src/crypto/frost.rs:488]: @jferrant here's a draft of the follow on PR with a results channel: 

https://github.com/stacks-network/stacks-blockchain/pull/3923

**Review by jferrant — APPROVED** [review]
Excited to see this merged. :)

File: stacks-signer/src/config.rs
Comment (jcnelson) [reply stacks-signer/src/config.rs:99]: Thanks for updating the comment @jferrant!

@xoloki I agree in principal that the keys ought to be separate and have different storage strategies, but I'm curious why the signer needs to have a key that controls large amounts of STX in the first place?  When the user stacks their STX in `pox-4`, they'd additionally set a dedicated signing key for the purposes of signing blocks, chunks, and transactions (i.e. the counterpart to `message_private_key`).  The STX would presumably be managed by the user's wallet, not the signer.  Additionally, the PoX rewards would be sent to a (separate) address of the user's choosing, which almost certainly would not be under the control of the signer.  Please help me understand what I'm missing?

File: stacks-signer/src/crypto/frost.rs
Comment (jferrant) [reply stacks-signer/src/crypto/frost.rs]: Oh missed this. Adding https://github.com/stacks-network/stacks-blockchain/issues/3924

File: libsigner/src/runloop.rs
Comment (jferrant) [inline libsigner/src/runloop.rs:70]: I think this actually can be fixed in this PR. If you just update the duration set in the timeout in the config file to be 128 millisecs, it should have the same effect. It is an optional that defaults to 5 seconds right now. however, you could add it as an option in the build_singer_config_tomls function or just chane the EVENT_TIMEOUT_SECS default.

File: stacks-signer/src/crypto/frost.rs
Comment (jcnelson) [reply stacks-signer/src/crypto/frost.rs:100]: Either way is fine.  If you do it in a subsequent PR, then can you create an issue for it?  Thanks!

File: stacks-signer/src/crypto/frost.rs
Comment (jcnelson) [inline stacks-signer/src/crypto/frost.rs:299]: I remember a while back we were discussing this possible outcome.  I don't remember this point, though: can this happen by accident, or does this condition only arise from malicious behavior on the coordinator's part?

Either way the signers need to re-run DKG, but if this only happens due to malicious behavior, then should the logic here identify the malicious actor and exclude them from the next DKG round?

Also, regarding the comment below -- does _not_ incrementing the `dkg_public_id` (or having some way of tracking the distribute/gather instance) mean that some signers' gathered public shares could be from a "stale" distribute/gather cycle?  If so, is that a problem?

File: stacks-signer/src/crypto/frost.rs
Comment (jcnelson) [reply stacks-signer/src/crypto/frost.rs:415]: Gotcha, thanks!

File: stacks-signer/src/crypto/frost.rs
Comment (jcnelson) [inline stacks-signer/src/crypto/frost.rs:447]: Would it make sense to use a draining iterator for the following?  This state won't be used again in this signing round, right?  It would save you some `clone()`s and ensure this signature generation step only happens at most once.

File: stacks-signer/src/runloop.rs
Comment (jcnelson) [reply stacks-signer/src/runloop.rs:31]: > There's lots of weird crypto things that can happen if you can trick people into signing arbitrary messages.

That's what I'm worried about.

> It's definitely something to think about but seems very orthogonal to this PR itself.

Either way, it seems important that we don't forget to do this.  Can you put this into an issue before merging?

File: stacks-signer/src/runloop.rs
Comment (jcnelson) [reply stacks-signer/src/runloop.rs:228]: Gotcha.  Just trying to make sure that they don't get dropped on the floor :pray:

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:101]: Linking the comment I had on the other PR here for visibility: https://github.com/stacks-network/stacks-blockchain/pull/3923#discussion_r1326362766

**Comment by jcnelson** [comment]
> BTW CI is failing with: @xoloki 

I get something similar when building on Alpine.  It seems that CI needs to be updated to install `libLLVM-16` development libraries and headers (in particular, the static libraries).  It comes from `p256k1`.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [reply stacks-signer/src/stacks_client.rs]: Sounds good; thanks for adding the issue @jferrant!

File: stacks-signer/src/crypto/frost.rs
Comment (jcnelson) [reply stacks-signer/src/crypto/frost.rs:299]: Ah nm, I see this was discussed in another comment thread :pray:

**Comment by jcnelson** [comment]
I only have three remaining concerns before approving:

* Getting CI to pass (probably just need to install LLVM dev packages)
* Trying to figure out why `stacks_private_key` is necessary
* Trying to figure out if we need more slots in the StackerDB, or if we can get away with giving the signer only one slot (instead of N slots for each of the N signing states).

File: stacks-signer/src/crypto/frost.rs
Comment (jferrant) [reply stacks-signer/src/crypto/frost.rs:100]: Issue made :D https://github.com/stacks-network/stacks-blockchain/issues/3925

File: stacks-signer/src/runloop.rs
Comment (jferrant) [reply stacks-signer/src/runloop.rs:31]: Issue created https://github.com/stacks-network/stacks-blockchain/issues/3926

File: stacks-signer/src/config.rs
Comment (xoloki) [reply stacks-signer/src/config.rs:100]: Totally possible that I'm making bad assumptions here.  I though the way PoX worked was that users stacked significant (~100k) STX for each reward slot, and that the STX address associated with the stacking wallets would be the same STX address that Signers will use to talk to sbtc contracts etc.

If this is a bad assumption, and the STX address the Signer will use with stackerDB is not the address which controls the large STX stake, then yeah we can just use the same private key for both.

**Comment by xoloki** [comment]
> * Getting CI to pass (probably just need to install LLVM dev packages)

I'll fix this today.

> * Trying to figure out why `stacks_private_key` is necessary

See my comment above.

> * Trying to figure out if we need more slots in the StackerDB, or if we can get away with giving the signer only one slot (instead of N slots for each of the N signing states).

We don't need a set of slots per key_id/reward slot, we just need a set of slots per stacking entity/Signer.  This will be more like 100, so with like 8 slots per `Signer` we only need like 800 slots total (currently we're using 16 slots per `Signer` but we don't need anywhere near that many).  I think we'll be okay.

File: stacks-signer/src/config.rs
Comment (jcnelson) [reply stacks-signer/src/config.rs:100]: The users don't stack to specific reward slots per se; they just stack a large quantity of STX, and depending on how much STX they stacked relative to everyone else, they are allotted zero or more reward slots by the system at the start of the next reward cycle.

That said, the key that holds the STX and performs the stacking operation would _not_ need to be known to the signer.  Instead, the act of stacking from this high-value key would assign a _different_ block-signing key for the signer to use.  Depending on what you think about this, it may even be desirable to have the user delegate _two_ keys to the signer: one for signing Stacks blocks, and one for signing Bitcoin sBTC transactions.  But in no case would the signer ever possess the key that controls the user's STX.

**Review by jcnelson — APPROVED** [review]
LGTM. Thank you for your clarifications @xoloki!  Once CI passes, feel free to merge :pray:

**Comment by xoloki** [comment]
Ugh, CI just keeps throwing more errors:

#11 598.3    Compiling stacks-node v0.1.0 (/build/testnet/stacks-node)
#11 653.7 LLVM ERROR: IO failure on output stream: No space left on device

**Pull request merged**
